Configure `query_constraints_list` to use `primary_key` by default

nvasilevski opened the pull request:
https://github.com/rails/rails/pull/46331 PR Introduced `query_constraints_list` configuration along with a backwards compatibility condition to keep models with no `query_constraints_list` configured to use `primary_key` as the only column for updates/deletes/reloads
https://github.com/Shopify/rails/blob/54cc42befd6bec0591ab064ec6e955131cdf8878/activerecord/lib/active_record/persistence.rb#L1064

This PR removes these backwards-compatibility checks in favor of an implicit configuration by convention that sets `query_constraints_list` to the model's `primary_key`

Just to serve as a spec, here is our expectations on which way of setting `query_constraints_list` should take precedence (from highest to lowest)
1. Manually configured `query_constraints_list` should have the highest precedence
2. If `primary_key` is manually configured, `query_constraints_list` should use the manually configured `primary_key` value
3. As the lowest priority, `query_constraints_list` should be using `primary_key` derived from schema

### Solution

`query_constraints_list` becomes a **private** getter for the newly introduced `@_query_constraints_list` instance variable and by calling `query_constraints_list_fallback ` sets `@_query_constraints_list` to either the `primary_key` if we are currently in a `base_class?` or if we are in a child class which redefined the `primary_key` and it's different from parent's `primary_key. Otherwise it falls back to the parent's `query_constraints_list`

### Alternative solution

At some point I was considering `primary_key.rb` to be the one responsible for configuring `query_constraints_list` but it doesn't feel right for a primary key to know about existence of `query_constraints` module. 
Another option would be to have an abstraction in the middle that knows about both `primary_key` and `query_constraints` and makes sure those values are properly configured but I don't see any need in such abstraction at the moment.

nvasilevski committed:
Configure `query_constraints_list` to use `primary_key` by default

Every `ActiveRecord::Base` model now should have `query_constraints_list`
configured by default and its value set to the model's `primary_key`

nvasilevski commented:
Well, the issue with a child model that uses different `primary_key` is the real one so I'm going to take some time trying to find a solution for a child model to inherit parent's `query_constraints_list` while respecting it's own `primary_key` if it's being explicitly overridden

nvasilevski commented:
I've updated the implementation and added more test cases to cover examples with inheritance

eileencodes commented on activerecord/lib/active_record/query_constraints.rb:
```diff
@@ -39,7 +35,35 @@ module ClassMethods
+      #   end
+      #
+      #   Comment.query_constraints_list # => ["post_id", "id"]
+      def query_constraints_list
```
I don't think think should be part of the public API (at least not yet), so let's no-doc this. I don't really love "query_constraints_list" as an API and don't think applications should ever need to list out their query constraints list. It feels like this was implemented so we can test it, but I'd rather test that the queries do what we want rather than have Models output their query constraints.

nvasilevski replied:
>It feels like this was implemented so we can test it

You are 100% right on this, I started from tests as I wanted to have a list of specs to make sure that inheritance works as we expect it and end up making the method public 🙃 
I'm going to change it to private and rewrite tests to assert for queries 👍

eileencodes commented on activerecord/lib/active_record/query_constraints.rb:
```diff
@@ -39,7 +35,35 @@ module ClassMethods
+      # For a base class, just use the primary key.
+      # For a child class, use the primary key unless primary key was overridden.
+      # If the child's primary key was not overridden, use the parent's query_constraints_list.
+      def query_constraints_list_fallback # :nodoc:
```
Let's put this under `private`.

eileencodes commented on activerecord/test/cases/query_constraints_test.rb:
```diff
@@ -11,4 +14,22 @@ def test_primary_key_stays_the_same
+  def test_child_keeps_parents_query_constraints
+    assert_equal(["clothing_type", "color"], ClothingItem.query_constraints_list)
+    assert_equal(["clothing_type", "color"], ClothingItem::Shirt.query_constraints_list)
+  end
```
I think we should remove these tests and only test the queries are correct.

nvasilevski commented:
Updated `query_constraints_list` to be private from rails api point of view. From ruby point of view it's public as we need to use it ourselves but we don't expect apps to rely on it. Also updated tests to check sql queries instead of `query_constraints_list` directly
